Source/WebCore/rendering/svg/RenderSVGInlineText.cpp

```diff
@@ -48,33 +49,18 @@ namespace WebCore {
 
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(RenderSVGInlineText);
 
-static String applySVGWhitespaceRules(const String& string, bool preserveWhiteSpace)
+static String normalizeWhitespace(const String& string)
 {
+    // Turn tabs, newlines and carriage returns into spaces.
     String newString = string;
-    if (preserveWhiteSpace) {
-        // Spec: When xml:space="preserve", the SVG user agent will do the following using a
-        // copy of the original character data content. It will convert all newline and tab
-        // characters into space characters. Then, it will draw all space characters, including
-        // leading, trailing and multiple contiguous space characters.
-        newString = makeStringByReplacingAll(newString, '\t', ' ');
-        newString = makeStringByReplacingAll(newString, '\n', ' ');
-        newString = makeStringByReplacingAll(newString, '\r', ' ');
-        return newString;
-    }
-
-    // Spec: When xml:space="default", the SVG user agent will do the following using a
-    // copy of the original character data content. First, it will remove all newline
-    // characters. Then it will convert all tab characters into space characters.
-    // Then, it will strip off all leading and trailing space characters.
-    // Then, all contiguous space characters will be consolidated.
-    newString = makeStringByReplacingAll(newString, '\n', ""_s);
-    newString = makeStringByReplacingAll(newString, '\r', ""_s);
     newString = makeStringByReplacingAll(newString, '\t', ' ');
+    newString = makeStringByReplacingAll(newString, '\n', ' ');
```

Comment:
This is really inefficient. We must have code somewhere to replace multiple characters in one traversal.

Is there an expectation that CRLF will get converted to a single space?

Reply:
We fail following test case across browser - https://jsfiddle.net/asyrzuz2/, we use to do `emptyString` (""_s) before.

I will search through code to see, if there is a way to do all these via single go.

Reply:
The `WTF::String::simplifyWhitespace` function is close to what we want. Especially because the comment mentions “all contiguous space characters will be consolidated” and that’s part of what `simplifyWhitespace` does. But it’s possible we’ll have to write a new function if we want something that’s super efficient and does exactly the right operation, and that is probably beyond the scope of what we should do in this pull request that is focused on correctness rather than efficiency. Scanning the string multiple times is not great, but doing so with the super-efficient “just check for one specific character” case might even be faster than scanning once with a less efficient inner loop.